protocol/src/types/primitive.rs

```diff
@@ -24,19 +25,23 @@ pub const METADATA_KEY: &str = "metadata";
 
 lazy_static! {
     static ref HASHER_INST: HasherKeccak = HasherKeccak::new();
-    static ref ADDRESS_HRP: ArcSwap<String> = ArcSwap::from(Arc::new("muta".to_owned()));
+    static ref ADDRESS_HRP: ArcSwap<SmolStr> = ArcSwap::from(Arc::new("muta".into()));
     static ref ADDRESS_HRP_INITED: AtomicBool = AtomicBool::new(false);
 }
 
-pub fn address_hrp() -> Arc<String> {
-    ADDRESS_HRP.load_full()
+pub fn address_hrp() -> SmolStr {
+    ADDRESS_HRP.load().as_ref().clone()
 }
 
 pub fn init_address_hrp(address_hrp: String) {
     if ADDRESS_HRP_INITED.load(Ordering::SeqCst) {
         panic!("address hrp can only be inited once");
     }
 
+    if address_hrp.as_bytes().len() > 21 {
```

Comment:
please use is_heap_allocated() instead of 21

And, do not restrict the bech32 standard
I reckon if the hrp is allocated on heap, just throw a warning of perf impacting